src/Compilers/Core/Portable/DiagnosticAnalyzer/ShadowCopyAnalyzerAssemblyLoader.cs

```diff
@@ -11,60 +11,69 @@
+        /// up the shadow directory.
+        /// </summary>
+        private readonly Mutex _mutex;
+
         internal readonly Task DeleteLeftoverDirectoriesTask;
 
+        private int _directoryCount;
+
+        private readonly ConcurrentDictionary<string, int> _directoryMap = new();
+
         /// <summary>
-        /// The directory where this instance of <see cref="ShadowCopyAnalyzerAssemblyLoader"/>
-        /// will shadow-copy assemblies, and the mutex created to mark that the owner of it is still active.
+        /// Map from the shadow assembly path to the original assembly path
         /// </summary>
-        private readonly Lazy<(string directory, Mutex)> _shadowCopyDirectoryAndMutex;
+        private readonly ConcurrentDictionary<string, string> _shadowToOriginalPathMap = new();
 
-        private readonly ConcurrentDictionary<Guid, Task<string>> _mvidPathMap = new ConcurrentDictionary<Guid, Task<string>>();
-        private readonly ConcurrentDictionary<(Guid, string), Task<string>> _mvidSatelliteAssemblyPathMap = new ConcurrentDictionary<(Guid, string), Task<string>>();
+        private readonly ConcurrentDictionary<string, Task> _copyMap = new();
 
         internal string BaseDirectory => _baseDirectory;
 
-        internal int CopyCount => _mvidPathMap.Count;
+        internal string ShadowDirectory => _shadowDirectory;
 
-#if NET
-        public ShadowCopyAnalyzerAssemblyLoader(string baseDirectory, ImmutableArray<IAnalyzerAssemblyResolver>? externalResolvers = null)
-            : this(null, baseDirectory, externalResolvers)
-        {
-        }
+        internal int CopyCount => _copyMap.Count;
 
-        public ShadowCopyAnalyzerAssemblyLoader(AssemblyLoadContext? compilerLoadContext, string baseDirectory, ImmutableArray<IAnalyzerAssemblyResolver>? externalResolvers = null)
-            : base(compilerLoadContext, AnalyzerLoadOption.LoadFromDisk, externalResolvers ?? [])
-#else
-        public ShadowCopyAnalyzerAssemblyLoader(string baseDirectory, ImmutableArray<IAnalyzerAssemblyResolver>? externalResolvers = null)
-            : base(externalResolvers ?? [])
-#endif
+        public ShadowCopyAnalyzerPathResolver(string baseDirectory)
         {
             if (baseDirectory is null)
             {
                 throw new ArgumentNullException(nameof(baseDirectory));
             }
 
             _baseDirectory = baseDirectory;
-            _shadowCopyDirectoryAndMutex = new Lazy<(string directory, Mutex)>(
-                () => CreateUniqueDirectoryForProcess(), LazyThreadSafetyMode.ExecutionAndPublication);
+            _shadowDirectoryName = Guid.NewGuid().ToString("N").ToLowerInvariant();
 
+            // The directory is deliberately _not_ created at this point. It will only be created when the first
+            // request comes in. This avoids creating unnecessary directories when no analyzers are loaded 
+            // via the shadow layer.
```

Comment:
i genuinely don't know: does this realistically happen?  might be good to hav ea comment if this is expected to be common enough to get this optimization.

Reply:
I actually tried removing this but the IDE depends on it. There are tests that cover this and they detail the IDE scenarios.